File: backend/src/models/Message.js
const mongoose = require('mongoose');
const { Schema, model } = mongoose;

// под-схема для прикреплённого файла к сообщению
const attachmentSchema = new Schema(
    {
        url: { type: String, required: true },
        type: {
            type: String,
            enum: ['image', 'file'],
            default: 'file',
        },
        name: { type: String },
        size: { type: Number }, // в байтах
    },
    { _id: false }
);

const messageSchema = new Schema(
    {
        room: {
            type: Schema.Types.ObjectId,
            ref: 'Room',
            required: true,
        },

        sender: {
            type: Schema.Types.ObjectId,
            ref: 'User',
            required: true,
        },

        text: {
            type: String,
            trim: true,
            default: '',
        },

        attachments: {
            type: [attachmentSchema],
            default: [],
        },

        // кто прочитал (список людей)
        readBy: {
            type: [{ type: Schema.Types.ObjectId, ref: 'User' }],
            default: [],
        },

        editedAt: {
            type: Date,
            default: null,
        },

        // soft delete — сообщение остаётся,
        // но помечается удалённым и текст на фронте заменяется на "сообщение удалено"
        isDeleted: {
            type: Boolean,
            default: false,
        },
    },
    { timestamps: true }
);

// главный индекс под историю чата: выборка по комнате + сортировка по дате
messageSchema.index({ room: 1, createdAt: -1 });

// курсорная пагинация у нас идёт по _id: room + _id < cursor
messageSchema.index({ room: 1, _id: -1 });

// на случай выборки "мои сообщения"
messageSchema.index({ sender: 1 });

module.exports = model('Message', messageSchema);

File: frontend/lib/features/message/data/models/message.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import '../../../user/data/models/user_model.dart';
import '../../domain/entities/message_entity.dart' show AttachmentType;

part 'message.freezed.dart';
part 'message.g.dart';

// соответствует enum'у type в Message.js: 'text' | 'system'
enum MessageType {
  @JsonValue('text')
  text,
  @JsonValue('system')
  system,
}

// соответствует enum'у systemData.action в Message.js
enum SystemAction {
  @JsonValue('participant_added')
  participantAdded,
  @JsonValue('participant_removed')
  participantRemoved,
  @JsonValue('participant_left')
  participantLeft,
  @JsonValue('participant_promoted')
  participantPromoted,
  @JsonValue('participant_demoted')
  participantDemoted,
}

// соответствует под-схеме systemData — присутствует только когда type == system
@freezed
class SystemData with _$SystemData {
  const factory SystemData({
    required SystemAction action,
    required PartialUser target,
  }) = _SystemData;

  factory SystemData.fromJson(Map<String, dynamic> json) =>
      _$SystemDataFromJson(json);
}

// под-схема для прикреплённого файла к сообщению
@freezed
class Attachment with _$Attachment {
  const factory Attachment({
    required String url,
    @Default(AttachmentType.file) AttachmentType type,
    String? name,
    int? size,
  }) = _Attachment;

  factory Attachment.fromJson(Map<String, dynamic> json) =>
      _$AttachmentFromJson(json);
}

@freezed
class Message with _$Message {
  const factory Message({
    required String id,
    required String room,
    required PartialUser sender, // populate('sender', 'username avatarUrl') — всегда объект
    @Default('') String text,
    @Default([]) List<Attachment> attachments,
    @Default([]) List<String> readBy,
    DateTime? editedAt,
    @Default(false) bool isDeleted,
    DateTime? createdAt,
    DateTime? updatedAt,
    @Default(MessageType.text) MessageType type,
    SystemData? systemData,
  }) = _Message;

  factory Message.fromJson(Map<String, dynamic> json) => _$MessageFromJson(json);
}